pkg/collector/corechecks/servicediscovery/discover_services_linux.go

```diff
@@ -0,0 +1,318 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+package servicediscovery
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/servicedetector"
+	"github.com/DataDog/datadog-agent/pkg/process/portlist"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+func init() {
+	newOSImpl = newLinuxImpl
+}
+
+type aliveMap struct {
+	m  map[int]*processInfo
+	mu sync.RWMutex
+}
+
+func (a *aliveMap) get(pid int) (*processInfo, bool) {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	v, ok := a.m[pid]
+	return v, ok
+}
+
+func (a *aliveMap) set(pid int, info *processInfo) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.m[pid] = info
+}
+
+func (a *aliveMap) delete(pid int) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	delete(a.m, pid)
+}
+
+type ignoreMap struct {
+	m  map[int]string
+	mu sync.RWMutex
+}
+
+func (i *ignoreMap) get(pid int) (string, bool) {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+	v, ok := i.m[pid]
+	return v, ok
+}
+
+func (i *ignoreMap) set(pid int, name string) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.m[pid] = name
+}
+
+func (i *ignoreMap) delete(pid int) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	delete(i.m, pid)
+}
+
+type linuxImpl struct {
+	sender    *telemetrySender
+	ignoreCfg map[string]struct{}
+
+	// PID -> process info
+	// pids that are considered services will be added here.
+	alive *aliveMap
+
+	// PID -> process name
+	// pids will be added here because the name was ignored in the config, or because
+	// it was already scanned and had no open ports.
+	ignore *ignoreMap
```

Comment:
I think we had discussed getting rid of the ignoreMap because a process that opens up a listen port after the scan will end up being ignore.